store-implementation/map-store/src/main/java/uk/gov/gchq/gaffer/mapstore/impl/AddElementsHandler.java

```diff
@@ -31,147 +32,123 @@
         final String group = element.getGroup();
         final Element elementWithGroupByProperties = element.emptyClone();
-        final Properties properties = new Properties();
-        groupToGroupByProperties.get(group)
-                .forEach(propertyName -> elementWithGroupByProperties
-                        .putProperty(propertyName, element.getProperty(propertyName)));
-        groupToNonGroupByProperties.get(group)
-                .forEach(propertyName -> properties.put(propertyName, element.getProperty(propertyName)));
-
-        Properties existingProperties = elementToProperties.get(elementWithGroupByProperties);
-        if (null == existingProperties) {
-            existingProperties = new Properties();
-            elementToProperties.put(elementWithGroupByProperties, existingProperties);
+        final GroupedProperties properties = new GroupedProperties(element.getGroup());
+        for (final String propertyName : mapImpl.getGroupByProperties(group)) {
+            elementWithGroupByProperties.putProperty(propertyName, element.getProperty(propertyName));
+        }
+        for (final String propertyName : mapImpl.getNonGroupByProperties(group)) {
+            properties.put(propertyName, element.getProperty(propertyName));
         }
 
-        schema.getElement(group).getIngestAggregator().apply(existingProperties, properties);
+        mapImpl.addAggElement(elementWithGroupByProperties, properties);
         return elementWithGroupByProperties;
     }
 
-    private Element updateElementToPropertiesNoGroupBy(final Element element,
-                                                       final Map<Element, Properties> elementToProperties) {
-        final Properties existingProperties = elementToProperties.get(element);
-        if (null == existingProperties) {
-            // Clone element and add to map with properties containing 1
-            final Element elementWithGroupByProperties = element.emptyClone();
-            elementWithGroupByProperties.copyProperties(element.getProperties());
-            final Properties properties = new Properties();
-            properties.put(MapImpl.COUNT, 1);
-            elementToProperties.put(elementWithGroupByProperties, properties);
+    private Element addNonAggElement(final Element element, final Schema schema, final MapImpl mapImpl) {
+        final Element elementClone = element.emptyClone();
+
+        // Copy properties that exist in the schema
+        final SchemaElementDefinition elementDef = schema.getElement(element.getGroup());
+        if (elementDef.getProperties().equals(element.getProperties().keySet())) {
+            elementClone.copyProperties(element.getProperties());
+        } else {
+            for (final String property : elementDef.getProperties()) {
+                elementClone.putProperty(property, element.getProperty(property));
+            }
+        }
```

Comment:
Is the test in the if necessary? The `elementClone.copyProperties` method just iterates through the provided properties and adds them. The for loop in the else code does the same thing. Either way the result is the same - only properties from `elementDef` get added (which is the right thing to do). Comparing two sets to see if they're the same is possibly a bit expensive (and this method will be called a lot).